src/OpportunisticResourceEstimator.cpp

```diff
@@ -0,0 +1,122 @@
+#include "OpportunisticResourceEstimator.hpp"
+#include "CommandRunner.hpp"
+#include "Helpers.hpp"
+
+#include <google/protobuf/text_format.h>
+
+#include <process/defer.hpp>
+#include <process/dispatch.hpp>
+
+namespace criteo {
+namespace mesos {
+
+using namespace process;
+using namespace mesos;
+
+using std::string;
+
+using ::mesos::modules::Module;
+using ::mesos::Resource;
+using ::mesos::Resources;
+using ::mesos::slave::ResourceEstimator;
+using ::mesos::ResourceUsage;
+
+using process::Future;
+using process::Failure;
+using process::terminate;
+
+class OpportunisticResourceEstimatorProcess
+    : public process::Process<OpportunisticResourceEstimatorProcess> {
+ public:
+  OpportunisticResourceEstimatorProcess(
+      const lambda::function<Future<ResourceUsage>()>& _usage,
+      const Option<Command>& oversubscribableCommand, bool isDebugMode);
+
+  virtual process::Future<Resources> oversubscribable();
+  virtual process::Future<Resources> _oversubscribable(
+      const ResourceUsage& usage);
+
+ protected:
+  const lambda::function<Future<ResourceUsage>()> usage;
+
+ private:
+  Option<Command> m_oversubscribableCommand;
+  bool m_isDebugMode;
+};
+
+OpportunisticResourceEstimatorProcess::OpportunisticResourceEstimatorProcess(
+    const lambda::function<process::Future<ResourceUsage>()>& _usage,
+    const Option<Command>& oversubscribableCommand, bool isDebugMode)
+    : usage(_usage),
+      m_oversubscribableCommand(oversubscribableCommand),
+      m_isDebugMode(isDebugMode) {}
+
+Future<Resources> OpportunisticResourceEstimatorProcess::_oversubscribable(
+    const ResourceUsage& usage) {
+  // Mocking a resources to have valid return
+  Try<Resources> _resources = ::mesos::Resources::parse(
+      "[{\"name\" : \"cpus\", \"type\":\"SCALAR\", \"scalar\" : {\"value\" : "
+      "\"0\"}, \"role\" : \"*\"}]");
+  Resources resources = _resources.get();
+
+  if (m_oversubscribableCommand.isNone()) {
+    return resources;
+  }
+  logging::Metadata metadata = {"0", "oversubscribable"};
+
+  CommandRunner cmdrunner = CommandRunner(m_isDebugMode, metadata);
+
+  string input;
+  google::protobuf::TextFormat::PrintToString(usage, &input);
+
+  Try<string> output = cmdrunner.run(m_oversubscribableCommand.get(), input);
+  if (output.isError()) {
+    return resources;
```

Comment:
I think we should:
- log the error
- return the error case of the Try class (unless mesos reacts weirdly to this)

Reply:
I have added a LOG(INFO) to retrieve the error
I am still investigating what i could give instead of resources

Reply:
ack